### PR TITLE
fix error with `plan(multisession)` when `parallel_over = "everything"`

### LINKED ISSUE #888: `plan(multisession)` errors with `parallel_over = "everything"`
When tuning with `tune_race_anova` with parallelism using `future::plan(multisession)`, I get `Error: object 'iteration' not found`.

``` r
library(tidymodels)
library(finetune)
library(discrim)
#> 
#> Attaching package: 'discrim'
#> The following object is masked from 'package:dials':
#> 
#>     smoothness
library(future)

data(two_class_dat, package = "modeldata")

set.seed(1)
rs <- bootstraps(two_class_dat, times = 10)

rda_spec <-
  discrim_regularized(frac_common_cov = tune(), frac_identity = tune()) %>%
  set_engine("klaR")

grid <-
  rda_spec %>%
  extract_parameter_set_dials() %>%
  grid_max_entropy(size = 20)

ctrl <- control_race(verbose_elim = TRUE)

set.seed(4)

plan(multisession)

grid_anova <- 
  rda_spec %>% 
  tune_race_anova(Class ~ ., resamples = rs, grid = grid, control = ctrl)
#> ℹ Evaluating against the initial 3 burn-in resamples.
#> Error: object 'iteration' not found
```

<sup>Created on 2024-04-10 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

With `plan(sequential)` I get no errors.

### PR DESCRIPTION
**Pull request by simonpcouch**
Closes #888. Opening as draft because _I do not like it_ but for some reason, with `plan(multisession)` in this context, foreach/doFuture is unable to find counters defined in the outer loop from the inner loop. Will come back to this in a few hours.

### CODE REVIEW

**Comment by simonpcouch** [comment]
The reprex from the linked issue, with this PR:

``` r
library(tidymodels)
library(future)

plan(multisession)

set.seed(1)
grid_res <- 
  tune_grid(
    logistic_reg(engine = "glmnet", penalty = tune()), 
    Class ~ ., 
    resamples = bootstraps(two_class_dat, times = 10), 
    control = control_grid(parallel_over = "everything")
  )
```

<sup>Created on 2024-04-17 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

Still feel a little sketched out by the RNG, but with a version of this PR that inlines `cat(seed); cat("\n")` in `tune_grid_loop_iter`:

``` r
library(tidymodels)
library(future)

set.seed(1)
grid_res <- 
  tune_grid(
    logistic_reg(engine = "glmnet", penalty = tune()), 
    Class ~ ., 
    resamples = bootstraps(two_class_dat, times = 10)
  )
#> 10407 -595255793 1999557892 -1504327627 1443955570 -868316085 1199646480
#> 10407 -766885743 588825389 -1237162594 -2077545013 1034854371 1346017440
#> 10407 643866970 -1425241749 167198522 -1402892000 2029063016 465008873
#> 10407 1491927262 308977226 1434171751 -1167638779 1315410366 -1216191782
#> 10407 453278401 -1412304971 -1981979732 -499884549 -1894324098 -1074235488
#> 10407 1976043951 1590119312 -275731398 172882762 -1632292343 714967610
#> 10407 1752334117 -1982243124 425861975 -2054719820 1986628774 -1068611118
#> 10407 1374489337 1779652776 -1593829796 1632076891 -1778750734 -1916375886
#> 10407 518174087 -81165699 1692243331 -1120474810 1861794411 1751884798
#> 10407 -1838125115 664378784 -415651163 -1719002853 1581429226 692082129

plan(multisession)

set.seed(1)
grid_res <- 
  tune_grid(
    logistic_reg(engine = "glmnet", penalty = tune()), 
    Class ~ ., 
    resamples = bootstraps(two_class_dat, times = 10), 
    control = control_grid(parallel_over = "everything")
  )
#> 10407 -595255793 1999557892 -1504327627 1443955570 -868316085 1199646480
#> 10407 -766885743 588825389 -1237162594 -2077545013 1034854371 1346017440
#> 10407 643866970 -1425241749 167198522 -1402892000 2029063016 465008873
#> 10407 1491927262 308977226 1434171751 -1167638779 1315410366 -1216191782
#> 10407 453278401 -1412304971 -1981979732 -499884549 -1894324098 -1074235488
#> 10407 1976043951 1590119312 -275731398 172882762 -1632292343 714967610
#> 10407 1752334117 -1982243124 425861975 -2054719820 1986628774 -1068611118
#> 10407 1374489337 1779652776 -1593829796 1632076891 -1778750734 -1916375886
#> 10407 518174087 -81165699 1692243331 -1120474810 1861794411 1751884798
#> 10407 -1838125115 664378784 -415651163 -1719002853 1581429226 692082129
```

<sup>Created on 2024-04-17 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

File: R/grid_code_paths.R [hunk 1]
```diff
@@ -225,6 +225,8 @@ tune_grid_loop_impl <- function(fn_tune_grid_loop_iter,
 
     seeds <- generate_seeds(rng, n_splits * n_grid_info)
 
+    slice_seeds <- slice_seeds
```
Comment (topepo) [inline R/grid_code_paths.R:228]: Yeah, this isn't great, but it's probably what needs to happen now. Perhaps @HenrikBengtsson has a suggestion for this.

Comment (simonpcouch) [reply R/grid_code_paths.R:228]: The other option here is [`.options.future = list(globals = structure(..., add = ...))`](https://github.com/HenrikBengtsson/doFuture/blob/95b8e80db6eeb4017eb6be645fe54c817b7dcbb7/R/dofuture_OP.R#L26-L34).

Comment (HenrikBengtsson) [reply R/grid_code_paths.R:228]: > The other option here is [`.options.future = list(globals = structure(..., add = ...))`](https://github.com/HenrikBengtsson/doFuture/blob/95b8e80db6eeb4017eb6be645fe54c817b7dcbb7/R/dofuture_OP.R#L26-L34).

Another way is to "guide" Futureverse (under the hood it's the **globals** package) by adding a dummy use of the global variable that it fails to detect, e.g.

```r
f <- future({ 
  b # "declare" 'b' as a global variable
  a * get("b") 
})
```

which effectively works the same as:

```r
f <- future({ 
  a * get("b") 
}, globals = structure(TRUE, add = "b"))
```

This is mentioned in https://future.futureverse.org/articles/future-4-issues.html#missing-globals-false-negatives

**Pull request merged**